Renamed some routing APIs to better express their unicast character.

SzymonPobiega opened the pull request:
Changes:
 * `busConfig.Routing()` -> `busConfig.UnicastRouting()`
 * `AddStatic` -> `Add`
 * Moved advanced routing features to `busConfig.AdvancedRouting`
 * Moved filed-based routing config to be available only in *advanced routing* and MSMQ transport config
 * Moved `Publishers` to *advanced routing* and exposed the most common methods directly on `UnicastRouting`.

The new non-advanced API (`busConfig.UnicastRouting()`) is following:

```
RouteToEndpoint(string endpoint);
AddPublisher(string publisherEndpoint, Type eventType);
AddPublisher(string publisherEndpoint, Assembly eventAssembly, string namespace);
```

[Doco PR](https://github.com/Particular/docs.particular.net/pull/1313)

andreasohlund commented:
I like it :+1: 


Will you add the


```
if(transport.IsMultiCast && user.HasConfiguredUnicastStuff)
{
    throw new WhatAreYouTryingToDoException();

}

```

in a different pull?


I can't come up with anything(perhaps @johnsimons can) but is there a way for us to only provide this API then the user has selected a unicast transport?

SzymonPobiega commented:
I've played with various APIs that would enforce unicast/multicast in compile time but failed because:
 * You can pass a `Type` into `UseTransport`
 * The transport might decide if it should use unicast or multicast based on settings

So I decided to rename the `Routing()` into `UnicastRouting()`

andreasohlund commented:
>So I decided to rename the Routing() into UnicastRouting()

Fair enough, much better than the current naming

danielmarbach commented on src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt:
```diff
@@ -2200,8 +2200,8 @@ namespace NServiceBus.Routing
-        public void AddDynamic(System.Func<NServiceBus.Routing.EndpointName, System.Threading.Tasks.Task<System.Collections.Generic.IEnumerable<NServiceBus.Routing.EndpointInstance>>> dynamicRule) { }
-        public void AddStatic(NServiceBus.Routing.EndpointName endpoint, params NServiceBus.Routing.EndpointInstance[] instances) { }
+        public void Add(NServiceBus.Routing.EndpointName endpoint, params NServiceBus.Routing.EndpointInstance[] instances) { }
+        public void AddExternalProvider(System.Func<NServiceBus.Routing.EndpointName, System.Threading.Tasks.Task<System.Collections.Generic.IEnumerable<NServiceBus.Routing.EndpointInstance>>> dynamicRule) { }
```
Isn't `Provider` a totally generic word without meaning? How about just call it `Add`, or `AddExternal` or `Add` and `AddDynamic`. Actually `AddDynamic` wasn't too bad

SzymonPobiega replied:
ok, I'll leave `AddDynamic`.

andreasohlund commented:
Needs a rebase and the doco pull

andreasohlund commented:
@danielmarbach added you as a maintainer since you're already involved

SzymonPobiega commented:
I'll get to that PR, eventually...

danielmarbach commented:
@SzymonPobiega I could do the rebase for you if you are busy

SzymonPobiega commented:
@danielmarbach reverted back to `AddDynamic` and added `RouteToEndpoint` directly on `UnicastRoutingSettings`.

andreasohlund commented:
Needs a rebase due to @danielmarbach cleaning frenzy :)

andreasohlund commented on src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/Publishers.cs:
```diff
@@ -12,37 +12,40 @@ public class Publishers
     {
         List<Rule> rules = new List<Rule>();
 
+        internal IEnumerable<PublisherAddress> GetPublisherFor(Type eventType)
```
Should be below the public methods, do a code cleanup?

danielmarbach replied:
We need a cleanup on save rule for all contributors

SzymonPobiega commented:
Hmmm should I do something here?

andreasohlund commented:
Rebase and a ctrl+E+C on Core/Routing/MessageDrivenSubscriptions/Publishers.cs ?

SzymonPobiega commented:
:+1:

SzymonPobiega commented:
Please re-review. I've made some changes after a call where we reviewed the API once again.

danielmarbach commented on src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs:
```diff
@@ -105,9 +104,7 @@ public SubscriberB()
-                    var publisher = new EndpointName(Conventions.EndpointNamingConvention(typeof(Publisher)));
-                    c.Publishers().AddStatic(publisher, typeof(MyEvent));
-                    c.Routing().EndpointInstances.AddStatic(publisher, new EndpointInstance(publisher));
+                    c.UnicastRouting().AddPublisher(PublisherEdndpoint, typeof(MyEvent));
```
Spelling

danielmarbach commented on src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs:
```diff
@@ -6,11 +6,12 @@
 
     public class When_publishing_to_scaled_out_subscribers_on_unicast_transports : NServiceBusAcceptanceTest
     {
+        static string PublisherEdndpoint => Conventions.EndpointNamingConvention(typeof(Publisher));
```
Spelling

danielmarbach replied:
@andreasohlund not ok to pull. Still has spelling things to fix

danielmarbach commented on src/NServiceBus.Core/Routing/UnicastRoutingSettings.cs:
```diff
@@ -0,0 +1,63 @@
+            GetOrCreate<Publishers>().Add(publisherEndpoint, eventAssembly, eventNamespace);
+        }
+
+        T GetOrCreate<T>()
```
Can't the routing feature just add a default instance of the publishers in the feature constructor?

SzymonPobiega replied:
It does but this API can be used from `EndpointConfiguration` so before features have chance to register their defaults.

danielmarbach commented:
I think a doco PR would really help. I failed to understand why one extension point is called `UnicastRouting` and the other one is called `AdvancedRouting`. The intent for me is not clear.

danielmarbach commented:
@andreasohlund your thoughts? I'm hesitant to pull this in since I fail to understand the API.

andreasohlund commented:
>I failed to understand why one extension point is called UnicastRouting and the other one is called AdvancedRouting. 

Unicast makes sense to me since this is not applicable to MultiCast routing. I'm also confused about the advanced part, is that for both multicast and unicast?

SzymonPobiega commented:
I've added the WIP doco pull. The `Advanced` part currently has only stuff useful for unicast routing. Would it be better to use `config.UnicastRouting().Advanced.SomeMethodOrProperty`?

andreasohlund commented:
>Would it be better to use config.UnicastRouting().Advanced.SomeMethodOrProperty?

Sounds better!

SzymonPobiega commented:
@andreasohlund moved `Advanced` under `Unicast`, rebased and squashed so that there is a commit with all the changes and a separate one only with API approval.

andreasohlund commented on src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt:
```diff
@@ -18,6 +18,13 @@ namespace NServiceBus
         Local = 0,
         Remote = 1,
     }
+    public class AdvancedRoutingSettings : NServiceBus.Configuration.AdvanceExtensibility.ExposeSettings
```
Should this be `AdvancedUnicastRoutingSettings`?

SzymonPobiega commented:
Updated `Advanced` -> `Mapping`

andreasohlund commented:
:+1: to merge, @danielmarbach ?

danielmarbach commented:
Needs a rebase.

SzymonPobiega commented:
@danielmarbach rebased

andreasohlund commented:
@SzymonPobiega can you fix the 2 spelling issues that @danielmarbach mentions so we can pull?

SzymonPobiega commented:
@andreasohlund done.
